### ISSUE TITLE
[groheondus] Installation failed with 404 when using username / password 

### ISSUE THREAD

**Issue by openPhiL** [openhab/openhab-addons#6296]
I have set up a Grohe Sense Guard at home and in my IOs App. 
I have Openhab 2.5m4 running and installed the binding in PaperUI.
I add my credentials to the Account thing. 

## Expected Behavior
The Thing should go "online".

## Current Behavior
```
2019-10-28 21:22:32.855 [home.event.ThingAddedEvent] - Thing 'groheondus:account:65d45eb3' has been added.
2019-10-28 21:22:32.863 [event.ThingStatusInfoEvent] - 'groheondus:account:65d45eb3' updated: INITIALIZING
2019-10-28 21:22:32.865 [hingStatusInfoChangedEvent] - 'groheondus:account:65d45eb3' changed from UNINITIALIZED to INITIALIZING
2019-10-28 21:22:32.869 [event.ThingStatusInfoEvent] - 'groheondus:account:65d45eb3' updated: UNKNOWN
2019-10-28 21:22:32.871 [hingStatusInfoChangedEvent] - 'groheondus:account:65d45eb3' changed from INITIALIZING to UNKNOWN
==> /var/log/openhab2/openhab.log <==
2019-10-28 21:22:33.415 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.initialize()' on 'org.openhab.binding.groheondus.internal.handler.GroheOndusAccountHandler@a9ac5a': Unknown response with code 404
java.lang.IllegalArgumentException: Unknown response with code 404
	at org.grohe.ondus.api.actions.LoginAction.lambda$getToken$0(LoginAction.java:25) ~[?:?]
	at java.util.Optional.orElseThrow(Optional.java:290) ~[?:?]
	at org.grohe.ondus.api.actions.LoginAction.getToken(LoginAction.java:24) ~[?:?]
	at org.grohe.ondus.api.OndusService.login(OndusService.java:82) ~[?:?]
	at org.grohe.ondus.api.OndusService.login(OndusService.java:34) ~[?:?]
	at org.openhab.binding.groheondus.internal.handler.GroheOndusAccountHandler.initialize(GroheOndusAccountHandler.java:145) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]

	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [139:org.openhab.core:2.5.0.M4]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:52) [139:org.openhab.core:2.5.0.M4]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
==> /var/log/openhab2/events.log <==
2019-10-28 21:22:33.435 [event.ThingStatusInfoEvent] - 'groheondus:account:65d45eb3' updated: UNINITIALIZED (HANDLER_INITIALIZING_ERROR): Unknown response with code 404
2019-10-28 21:22:33.437 [hingStatusInfoChangedEvent] - 'groheondus:account:65d45eb3' changed from UNKNOWN to UNINITIALIZED (HANDLER_INITIALIZING_ERROR): Unknown response with code 404
==> /var/log/openhab2/openhab.log <==
2019-10-28 21:22:33.439 [ERROR] [core.thing.internal.ThingManagerImpl] - Exception occurred while initializing handler of thing 'groheondus:account:65d45eb3': Unknown response with code 404
java.lang.IllegalArgumentException: Unknown response with code 404
	at org.grohe.ondus.api.actions.LoginAction.lambda$getToken$0(LoginAction.java:25) ~[?:?]
	at java.util.Optional.orElseThrow(Optional.java:290) ~[?:?]
	at org.grohe.ondus.api.actions.LoginAction.getToken(LoginAction.java:24) ~[?:?]
	at org.grohe.ondus.api.OndusService.login(OndusService.java:82) ~[?:?]
	at org.grohe.ondus.api.OndusService.login(OndusService.java:34) ~[?:?]
	at org.openhab.binding.groheondus.internal.handler.GroheOndusAccountHandler.initialize(GroheOndusAccountHandler.java:145) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [139:org.openhab.core:2.5.0.M4]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:52) [139:org.openhab.core:2.5.0.M4]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

## Possible solution
Update the docu and setup of the binding to require a token, it look OK with a token - although the "thing" for my Sense is not working and remains UNINITIALIZED. I creata another Issue for that.

## Steps to Reproduce (for Bugs)
It is a fresh install, everyone should have this problem?

## Your Environment
Fresh install of OpenHab 2.5m4 on a virtual ubuntu linux with raspberian.

**Comment by FlorianSW** [comment]
Hi, sorry for the late reply :(

I tried to reproduce your bug by doing the following steps:

# Started an openHAB 2.5.0.M4 release using the official docker hub image
# Added the GROHE ONDUS binding through the paper UI
# Configured a new Account thing for this binding with a set of username/password

The thing got online after saving it :/ Could you please give me the information when you registered for this GROHE account? It seems that, after a specific date all new registered accounts only work with the OAuth based authentication, and not with a plain username/password login anymore, the error message would be what you posted :/ In that case, you would need to use the refresh_token authentication mechanism, instead of the username/password one.

I also checked the refresh_token authentication with the M4, which works as well. Once you gave the info when you registered the account, we can either close this bug or investigate deeper :)

**Comment by openPhiL** [comment]
Hi Florian, 
So, weekend tüftel time -  I did the same, I have a pretty fresh OpenHab 2.5 M4 system (VM Ubuntu). I have removed the things and the binding and restarted openhab. 
```
10-Nov-2019 18:01:21.085 [DEBUG] [org.openhab.binding.groheondus                    ] - ServiceEvent UNREGISTERING - {org.eclipse.smarthome.config.discovery.DiscoveryService}={service.id=460, service.bundleid=267, service.scope=singleton} - org.openhab.binding.groheondus
10-Nov-2019 18:01:32.304 [DEBUG] [org.openhab.binding.groheondus                    ] - ServiceEvent UNREGISTERING - {org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory}={service.id=459, service.bundleid=267, service.scope=bundle, component.name=org.openhab.binding.groheondus.internal.handler.GroheOndusHandlerFactory, component.id=294} - org.openhab.binding.groheondus
10-Nov-2019 18:01:32.308 [DEBUG] [org.openhab.binding.groheondus                    ] - ServiceEvent UNREGISTERING - {org.openhab.binding.groheondus.internal.AccountsServlet}={service.id=457, service.bundleid=267, service.scope=bundle, component.name=org.openhab.binding.groheondus.internal.AccountsServlet, component.id=293} - org.openhab.binding.groheondus
10-Nov-2019 18:01:32.313 [DEBUG] [org.openhab.binding.groheondus                    ] - ServiceEvent UNREGISTERING - {javax.servlet.ServletContext}={osgi.web.version=2.5.0.M4, osgi.web.contextpath=/, service.id=458, osgi.web.symbolicname=org.openhab.binding.groheondus, service.bundleid=267, service.scope=singleton, osgi.web.contextname=default} - org.openhab.binding.groheondus
10-Nov-2019 18:01:32.330 [DEBUG] [org.openhab.binding.groheondus                    ] - BundleEvent STOPPING - org.openhab.binding.groheondus
10-Nov-2019 18:01:32.331 [DEBUG] [org.openhab.binding.groheondus                    ] - BundleEvent STOPPED - org.openhab.binding.groheondus
10-Nov-2019 18:01:32.347 [DEBUG] [org.openhab.binding.groheondus                    ] - BundleEvent UNRESOLVED - org.openhab.binding.groheondus
10-Nov-2019 18:01:32.350 [DEBUG] [org.openhab.binding.groheondus                    ] - BundleEvent UNINSTALLED - org.openhab.binding.groheondus
```

Then, I installed the binding using PaperUI and created the account thing - that immediately received status Status: UNINITIALIZED - HANDLER_INITIALIZING_ERROR Unknown response with code 404 
```
10-Nov-2019 20:05:53.636 [DEBUG] [org.openhab.binding.groheondus                    ] - BundleEvent INSTALLED - org.openhab.binding.groheondus
10-Nov-2019 20:05:53.897 [DEBUG] [org.openhab.binding.groheondus                    ] - BundleEvent RESOLVED - org.openhab.binding.groheondus
10-Nov-2019 20:05:54.033 [DEBUG] [org.openhab.binding.groheondus                    ] - BundleEvent STARTING - org.openhab.binding.groheondus
10-Nov-2019 20:05:54.127 [DEBUG] [org.openhab.binding.groheondus                    ] - ServiceEvent REGISTERED - {javax.servlet.ServletContext}={osgi.web.version=2.5.0.M4, osgi.web.contextpath=/, service.id=460, osgi.web.symbolicname=org.openhab.binding.groheondus, service.bundleid=271, service.scope=singleton, osgi.web.contextname=default} - org.openhab.binding.groheondus
10-Nov-2019 20:05:54.134 [DEBUG] [org.openhab.binding.groheondus                    ] - ServiceEvent REGISTERED - {org.openhab.binding.groheondus.internal.AccountsServlet}={service.id=459, service.bundleid=271, service.scope=bundle, component.name=org.openhab.binding.groheondus.internal.AccountsServlet, component.id=293} - org.openhab.binding.groheondus
10-Nov-2019 20:05:54.170 [DEBUG] [org.openhab.binding.groheondus                    ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory}={service.id=461, service.bundleid=271, service.scope=bundle, component.name=org.openhab.binding.groheondus.internal.handler.GroheOndusHandlerFactory, component.id=294} - org.openhab.binding.groheondus
10-Nov-2019 20:05:54.171 [DEBUG] [org.openhab.binding.groheondus                    ] - BundleEvent STARTED - org.openhab.binding.groheondus
10-Nov-2019 20:06:19.310 [DEBUG] [ndus.internal.discovery.GroheOndusDiscoveryService] - initialize discovery service
10-Nov-2019 20:06:19.314 [DEBUG] [org.openhab.binding.groheondus                    ] - ServiceEvent REGISTERED - {org.eclipse.smarthome.config.discovery.DiscoveryService}={service.id=462, service.bundleid=271, service.scope=singleton} - org.openhab.binding.groheondus

```
(really nothing in the logs, I double checked...)
Next, I created the token and added it to the http://openhab2/groheondus 
and the status changed to:
Status: OFFLINE - CONFIGURATION_ERROR Login failed
```
10-Nov-2019 20:10:17.874 [DEBUG] [oheondus.internal.handler.GroheOndusAccountHandler] - Grohe api login failed
javax.security.auth.login.LoginException: 401 - Unauthorized
	at org.grohe.ondus.api.actions.RefreshTokenAction.refresh(RefreshTokenAction.java:23) ~[271:org.openhab.binding.groheondus:2.5.0.M4]
	at org.grohe.ondus.api.OndusService.login(OndusService.java:91) ~[271:org.openhab.binding.groheondus:2.5.0.M4]
	at org.grohe.ondus.api.OndusService.login(OndusService.java:51) ~[271:org.openhab.binding.groheondus:2.5.0.M4]
	at org.openhab.binding.groheondus.internal.handler.GroheOndusAccountHandler.initialize(GroheOndusAccountHandler.java:142) [271:org.openhab.binding.groheondus:2.5.0.M4]
	at org.openhab.binding.groheondus.internal.handler.GroheOndusAccountHandler.setRefreshToken(GroheOndusAccountHandler.java:80) [271:org.openhab.binding.groheondus:2.5.0.M4]
	at org.openhab.binding.groheondus.internal.AccountServlet.doPost(AccountServlet.java:111) [271:org.openhab.binding.groheondus:2.5.0.M4]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707) [31:javax.servlet-api:3.1.0]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) [31:javax.servlet-api:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:873) [90:org.eclipse.jetty.servlet:9.4.18.v20190429]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:542) [90:org.eclipse.jetty.servlet:9.4.18.v20190429]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71) [194:org.ops4j.pax.web.pax-web-jetty:7.2.10]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548) [87:org.eclipse.jetty.security:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1700) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:293) [194:org.ops4j.pax.web.pax-web-jetty:7.2.10]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480) [90:org.eclipse.jetty.servlet:9.4.18.v20190429]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1667) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80) [194:org.ops4j.pax.web.pax-web-jetty:7.2.10]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.Server.handle(Server.java:505) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305) [80:org.eclipse.jetty.io:9.4.18.v20190429]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) [80:org.eclipse.jetty.io:9.4.18.v20190429]
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117) [80:org.eclipse.jetty.io:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```
And this is where I am right now - any suggestions what I could do differently? 
One hint: I changed the password using the link you provided https://idp2-apigw.cloud.grohe.com/v3/iot/oidc/login  - And I remember having Issues with the logins.. is it possible that there are two credentials necessary - one for the account and one for the app or something like that? 

I just checked this theory:  When I want to change my pw in the APP, it askes for old password and new password. the old password, the one I have just set using the link of yours, is not recognized as the "old" password from within the app. Those are not identical. 
So I reset within the app and within your link to the same password - but no success - I cannot get passed the 401 - Unauthorized.

**Comment by openPhiL** [comment]
Ups, I had just sent the response and "then" recognized that I have used the wrong token dammit. 

So, at least I know the logs are working while I debugged that.. so next in line was adding the senseGuard - was auto-discovered. I added it, it became status 
Status: UNKNOWN

And in the logs, the only entry is: 
```
10-Nov-2019 20:36:20.847 [DEBUG] [ondus.internal.handler.GroheOndusSenseGuardHandler] - Invalid command received for channel. Expected OnOffType, received org.eclipse.smarthome.core.types.RefreshType.
```
(I use simpleMode, so the items are created automatically). 

I have not installed the device yet, no water is going through - Might this be a problem?
